src/termdag.rs

```diff
@@ -4,15 +4,25 @@ use crate::{
     Symbol,
 };
 
+/// Like [`Expr`]s but with sharing and deduplication.
+///
+/// Terms refer to their children indirectly as indexes into an ambient [`TermDag`].
```

Comment:
Would be best if we could prevent users from constructing these but still allow pattern-matching on them. Not sure how to